text/0000-raw-pointer-comparisons.md

```diff
@@ -0,0 +1,58 @@
+- Feature Name: raw-pointer-comparisons
+- Start Date: 2015-05-27
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Allow equality, but not order, comparisons between fat raw pointers
+of the same type.
+
+# Motivation
+
+Currently, fat raw pointers can't be compared via either PartialEq or
+PartialOrd (currently this causes an ICE). It seems to me that a primitive
+type like a fat raw pointer should implement equality in some way.
+
+However, there doesn't seem to be a sensible way to order raw fat pointers
+unless we take vtable addresses into account, which is relatively weird.
+
+# Detailed design
+
+Implement PartialEq/Eq for fat raw pointers, defined as comparing both the
+unsize-info and the address. This means that these are true:
+
+```Rust
+    &s as &fmt::Debug as *const _ == &s as &fmt::Debug as *const _ // of course
+    &s.first_field as &fmt::Debug as *const _
+        != &s as &fmt::Debug as *const _ // these are *different* (one
+	                                 // prints only the first field,
+					 // the other prints all fields).
+```
+
+But
+```Rust
+    &s.first_field as &fmt::Debug as *const _ as *const () ==
+        &s as &fmt::Debug as *const _ as *const () // addresses are equal
+```
+
+# Drawbacks
+
+Order comparisons may be useful for putting fat raw pointers into
+ordering-based data structures (e.g. BinaryTree).
+
+# Alternatives
+
+@nrc suggested to implement heterogeneous comparisons between all thin
+raw pointers and all fat raw pointers. I don't like this because equality
+between fat raw pointers of different traits is false most of the
+time (unless one of the traits is a supertrait of the other and/or the
+only difference is in free lifetimes), and anyway you can always compare
```

Comment:
Depending on your version of equality there are other ways this can be true. If you ignore the vtables, and only compare the pointers (which is how I would define fat-ptr equality) then you can have two fat pointers of unrelated types which are equal:

```
let x = &T;
let a: &Tr1 = x;
let b: &Tr2 = x;
a == b
```

As long as `T` implements both `Tr1` and `Tr2`, then the two traits do not have to be related in any way.

We haven't yet implemented casting from `&Tr1` to `&Tr1 + Tr2`, so any cast would have to be to `*u8` or something.